hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/parse/QueryConverter.java

```diff
@@ -156,40 +162,23 @@ private HazelcastSqlToRelConverter createSqlToRelConverter() {
      *  It is used instead of "expand" flag due to bugs in Calcite (see {@link #EXPAND}).
      * </li>
      * <li>
-     *  Transformation of distinct UNION to UNION ALL, merging the neighboring UNION relations.
-     * </li>
-     * <li>
      *  Check, if the relation uses cyclic user types, and if they are allowed - skip this step.
      * </li>
-     *
      * </ul>
      *
      * @param rel                       Initial relation.
      * @param cyclicUserTypesAreAllowed Whether cyclic user types are allowed.
      * @return Resulting relation.
+     * @implNote we're using Calcite's {@link HepProgramBuilder} to GC the rules we don't need earlier.
      */
     private static RelNode performUnconditionalRewrites(RelNode rel, boolean cyclicUserTypesAreAllowed) {
-        HepProgramBuilder hepProgramBuilder = new HepProgramBuilder();
-
-        // Correlated subqueries elimination rules
-        hepProgramBuilder.addRuleInstance(CoreRules.FILTER_SUB_QUERY_TO_CORRELATE);
-        hepProgramBuilder.addRuleInstance(CoreRules.PROJECT_SUB_QUERY_TO_CORRELATE);
-        hepProgramBuilder.addRuleInstance(CoreRules.JOIN_SUB_QUERY_TO_CORRELATE);
-
-        // Union optimization rules
-        hepProgramBuilder.addRuleInstance(CoreRules.UNION_MERGE);
-        hepProgramBuilder.addRuleInstance(CoreRules.UNION_TO_DISTINCT);
-
-        // Other rules
-        hepProgramBuilder.addRuleInstance(ExtractUpdateExpressionsRule.INSTANCE);
-        // Note: the way to check if cyclic type are allowed in SqlValidator is error-prone,
-        //  and the best way is to check it on pre-optimization phase.
-        if (!cyclicUserTypesAreAllowed) {
-            hepProgramBuilder.addRuleInstance(ScanCyclicTypeMustNotExecuteRule.INSTANCE);
+        // Note: 'cyclicUserTypesAreAllowed' is cluster-wise property, no changes in runtime are expected.
+        if (hepSubqueryRewriterProgram == null) {
```

Comment:
you do not care about visibility (no volatile) and multiple concurrent initialisations, but this is probably fine. you might however add a comment that this is conscious decision

Reply:
> you might however add a comment that this is conscious decision
Exactly. Plan reads are cheaper than volatile reads, so in long-term it is more optimal to allow a few multiple concurrent initialisations, but later have cheap plain reads.